### PR TITLE
New attempt to update bokehjs from pop_colors_and_alpha to _pop_visuals

### LINKED ISSUE #9444: Conversion from _pop_colors_and_alpha to _pop_visuals missing bokehjs
Ni! @bryevdv 

For good reasons, #9153 has been merged without the corresponding changes to bokehjs.

This issue tracks the progress to update bokehjs.

### PR DESCRIPTION
**Pull request by solstag**
This continues #9153. Still sorting out bugs, I'll probably need help with TS usage.

- [ ] issues: fixes #9444
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

### CODE REVIEW

**Comment by mattpap** [comment]
@bryevdv, can we disable those bot messages?

**Comment by bryevdv** [comment]
Not without turning off the tool AFAIK. There is some useful actionable info here:

https://lgtm.com/projects/g/bokeh/bokeh/alerts/?mode=list&severity=

As well as some fans positives. There are ways to flag those, though, so I'd really prefer to try and see if the tool can be tailored for our codebase before ditching it.

**Comment by mattpap** [comment]
> Not without turning off the tool AFAIK.

Those bot comments are obtrusive enough to disturb flow of discussion, and I don't want to imagine how bad this will be in large PRs. This is clearly a bad design, especially that github has a spot on the page where a "check" can report status. As to overall usefulness of the tool, I'm happy to give it a chance, as long as it doesn't interfere with my work. However, looking at some results, there is plenty of duplication with our codebase tests and some false positives.

**Comment by bryevdv** [comment]
I found some more extensive docs, there may be more configuration. In any case I want some time to experiment with it. So we can timebox this until next week's Wednesday meeting and discuss removing it then, then after actually having some time to interact with it.

**Comment by solstag** [comment]
About the errors I'm getting, I've solved all but those like:

> tests/test_examples.py::test_js_examples[/home/travis/build/bokeh/bokeh/bokehjs/examples/categorical/categorical.html-example2-config2] 
> [JS] Error: property Rect.color wasn't declared

Which I can't figure out. I should of course set up debugging typescript locally (pointers are welcome), but if anyone has an idea of why this is happening...

In principle "color" should not appear in `result` because it is not a prop of Rect (line 658) and should be removed from `props` for the same reason plus the fact that "line_color" is a prop of Rect (lines 672 then 677). But for some reason it is leaking and giving that error.

**Comment by bryevdv** [comment]
@solstag I have the branch locally and can reproduce I will try to take a closer look tonight. In the mean time I think it would be good to get this up to date with master, that numpy dtype issue in some of the tests should go away.

**Comment by bryevdv** [comment]
@solstag  FYI there are also some linter errors: https://travis-ci.org/bokeh/bokeh/jobs/615781351?utm_medium=notification&utm_source=github_status It looks like running `node make lint --fix` will fix most if not all of them

**Comment by bryevdv** [comment]
@mattpap I see this locally either on this branch or on master
```

  should allow Paragraph
  should allow PreText
Rows of widgets
  should allow different content and fixed height
Tabs
  should allow tabs header location above
Cannot read property 'value' of null
[15:31:19] failed: Error: tests failed
[15:31:19]     at ChildProcess.proc.on (/Users/bryan/work/bokeh/bokehjs/make/tasks/test.ts:92:16)
[15:31:19]     at ChildProcess.emit (events.js:182:13)
[15:31:19]     at ChildProcess.EventEmitter.emit (domain.js:442:20)
[15:31:19]     at Process.ChildProcess._handle.onexit (internal/child_process.js:240:12)
```
Any ideas?

File: bokehjs/src/lib/api/plotting.ts [hunk 1]
```diff
@@ -630,30 +630,51 @@ export class Figure extends Plot {
...
+      if (_is_visual(pname)) {
+        let trait = _split_feature_trait(pname)[1]
+        if (props.hasOwnProperty(prefix+pname)) {
+            result[pname] = props[prefix+pname]
+            delete props[prefix+pname] }
+        else if (!cls.prototype.props.hasOwnProperty(trait)
+                 && props.hasOwnProperty(prefix+trait)) {
+            result[pname] = props[prefix+trait] }
+        else if (defaults.hasOwnProperty(pname)) {
+            result[pname] = defaults[pname] }
```
Comment (mattpap) [inline bokehjs/src/lib/api/plotting.ts]: Let's start from fixing this syntax (indentation, curly brackets, spacing, etc.), then we can get to fixing type errors.

**Comment by mattpap** [comment]
> Cannot read property 'value' of null

Most likely a bug in the test runner. Incidentally I'm currently working on improving this.

**Comment by bryevdv** [comment]
OK it actually seems to be the case that this error on the "minimal" test:
```
AttributeError: module 'numpy' has no attribute 'dtype'
```
is happening on every PR from an outside fork. I have no idea why that should, be, it appears to be installing the expected version of numpy... also GH CI does not appear to be running on external forks either. Will make a separate issue for these. grrr

**Comment by mattpap** [comment]
> also GH CI does not appear to be running on external forks either.

I suppose you will have to test those changes from both of your accounts, assuming the other doesn't have a privileged status.

**Comment by bryevdv** [comment]
> I suppose you will have to test those changes from both of your accounts, assuming the other doesn't have a privileged status. 

working with @ammarnajjar on this I think he will submit a PR which will also be the test case

**Comment by ammarnajjar** [comment]
> also GH CI does not appear to be running on external forks either.

in order for the actions to run on a fork, they must be enabled on the fork account first:
[docs](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows#pull-request-events-for-forked-repositories):

> Workflows don't run on forked repositories by default. You must enable GitHub Actions in the Actions tab of the forked repository.

**Comment by bryevdv** [comment]
@ammarnajjar But also this:

> When you create a pull request from a forked repository to the base repository, GitHub sends the pull_request event to the base repository and no pull request events occur on the forked repository.

We don't actually want (or at least don't care) if the actions run *in* the forked repo. We want them to run *in* this repo, when PRs from forks come in. I think your original suggestion of `pull_request` (or maybe `[push, pull_request]` even) is the correct solution and that we should try it in any case.

File: bokehjs/src/lib/api/plotting.ts (hunk 1 above)
Comment (solstag) [reply bokehjs/src/lib/api/plotting.ts]: ok, i guess it doesn't hurt to install nodejs on my pc. linting fixes committed.

**Comment by bryevdv** [comment]
@solstag @mattpap This block is not being compiled as expected:

```
for (const trait of traits) {
      delete props[prefix+trait] 
}
```
It's getting rendered as:
```
for (var _i = 0, traits_1 = traits; _i < traits_1.length; _i++) {
    var trait = traits_1[_i];
    delete props[prefix + trait];
}
```
which never executes the block, since `traits.length` is `undefined`

**Comment by mattpap** [comment]
> `for (const trait of traits) {`

This should have failed compilation. TypeScript will generate this optimized code only when `traits` is an array.

**Comment by solstag** [comment]
Hi all! Thanks for the exaplanations and, just to be sure, I get from them that I no longer have to change the code. Right? Otherwise I'd be fine replacing it with a different syntax.

**Comment by bryevdv** [comment]
@solstag Sorry for being unclear, there is still a code change needed (only one I think). This loop:

```
for (const trait of traits) {
      delete props[prefix+trait] 
}
```
Needs to be replaced by a formulation with `Set.forEach` When @mattpap says the other PR fixes this I believe he means that the TS compiler would throw an error (which I am not sure why it is not doing now)

**Comment by solstag** [comment]
> Needs to be replaced by a formulation with `Set.forEach` When @mattpap says the other PR fixes this I believe he means that the TS compiler would start to throw an error (which I am not sure why it is not doing now)

Oh, I thought he meant that upon moving to ES2017 this would work as is, since IE gets dropped, which is the only browser not implementing `for...of`. Anyway, if this makes the code more portable, I will apply the change.

**Comment by bryevdv** [comment]
@solstag this seems very close, do we need to mirror the `override_defaults` added to 

https://github.com/bokeh/bokeh/blob/master/bokeh/plotting/helpers.py#L316

?

**Comment by solstag** [comment]
You're right, this needs to be adapted to follow the new "override_defaults" behaviour.

I'm busy now, but will work on it next week. :)

**Comment by solstag** [comment]
Ni! @bryevdv it would seem there are no longer any PR related errors in the tests. Should be good to merge? Btw: happy and healthy 2020!

**Comment by bryevdv** [comment]
@solstag thanks for the kind words, and the persistence to see all this through. You are correct the only failure is unrelated and due to the examples report failing to upload (which is expected on PRs from forks, but should not fail the test, I will make a separate issue to follow up on that).

**Pull request merged**